### PR TITLE
Openvpn

### PR DESCRIPTION
**Pull request by jfelten**
Please review

### CODE REVIEW

File: incubator/openvpn/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
+description: A Helm chart for Kubernetes
+name: openvpn
+version: 0.1.0
```
Comment (coffeepac) [inline incubator/openvpn/Chart.yaml]: is this version 0.1.0?   or is this a 1.0.0?  Are there other features you want to add before considering it complete?

File: incubator/openvpn/Chart.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
+description: A Helm chart for Kubernetes
```
Comment (coffeepac) [inline incubator/openvpn/Chart.yaml]: please update to not boilerplate

File: incubator/openvpn/Chart.yaml (hunk 1 above)
Comment (coffeepac) [inline incubator/openvpn/Chart.yaml]: Can you please include the following fields:
- keywords (comma separated array of search terms.  I would expect openvpn and vpn at a minimum
- home  (URI for the openvpn project)
- maintainers (list of hashes where the hash keys are name and email.  should be you for now.)

File: incubator/openvpn/README.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# Helm chart for OpenVPN
+This chart will install an openvpn server inside a kubernetes cluster.  New certificates are generated on install, and a script is provided to genreate client keys as needed.  The chart will automatically configure dns to use kubeDNS and route all network traffic to kuberentes pods and services through the vpn.  By connecting to this VPN a host is effectively inside a cluster's network.
+
+###Uses
+The primary purpose of this chart was to make it easy to access kubernetes services during development.  It could also be used for any service that only needs to be access through a VPN, or as a general prupose VPN.
+
+##Usage
+
+		helm repo add incubator http://storage.googleapis.com/kubernetes-charts-incubator
```
Comment (coffeepac) [inline incubator/openvpn/README.md]: would prefer for the k2-charts version that we use the k2-charts repo information.  ie atlas.cnct.io

if you have no intention of merging this to k2-charts at all then disregard.

File: incubator/openvpn/README.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+Import the .ovpn file into your favorite openvpn tool like tunnelblick, and verify connectivity.
+
+##Configuration
+
+All configuration is in the values.yaml file, and can be overwritten via the helm --set flag.
+
+###Certificates
+
+Certificates are generated with each deployment.  This way there is no need to store them, but this may be an issue if the pods are restarted often or if there are many clients.  We may able a persistent storage of certs if this becomes an issue.
```
Comment (coffeepac) [inline incubator/openvpn/README.md]: should 'We may able' be 'We may enable'?

File: incubator/openvpn/README.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+###Important values
+* service.externalPort: 443 - external LoadBalancer port
+* service.internalPort: 443 - port of openVPN port
+* openvpn.OVPN_NETWORK: 10.240.0.0 - Network allocated for openvpn clients (default: 10.240.0.0).
+* openvpn.OVPN_SUBNET:  255.255.0.0 - Network subnet allocated for openvpn client (default: 255.255.0.0).
+* openvpn.OVPN_PROTO: tcp - Protocol used by openvpn tcp or udp (default: tcp).
+* openvpn.OVPN_K8S_POD_NETWORK: "10.0.0.0" - Kubernetes pod network (optional).
+* openvpn.OVPN_K8S_POD_SUBNET: "255.0.0.0" - Kubernetes pod network subnet (optional).
+
+####Note: As configured the chart will create a route for a large 10.0.0.0/8 network that may cause issues if that is your local network.  If so tweak this value to something more restrictive.  This route is added, because kubernetes genreate pods with IPs in this range.
```
Comment (coffeepac) [inline incubator/openvpn/README.md]: last sentence, no comma.

generate not genreate.

File: incubator/openvpn/mycert.ovpn [hunk 6]
```diff
@@ -0,0 +1,7 @@
+
```
Comment (coffeepac) [inline incubator/openvpn/mycert.ovpn]: What is this file for?

File: incubator/openvpn/templates/NOTES.txt [hunk 7]
```diff
@@ -0,0 +1,13 @@
+OpenVPN now Running.
```
Comment (coffeepac) [inline incubator/openvpn/templates/NOTES.txt]: is this file used anywhere?  It seems the information has already been put into the top level README.

File: incubator/openvpn/values.yaml [hunk 8]
```diff
@@ -0,0 +1,31 @@
+# Default values for openvpn.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+replicaCount: 1
+image:
+  repository: offlinehacker/openvpn-k8s
+  tag: latest
```
Comment (coffeepac) [inline incubator/openvpn/values.yaml]: this seems not great.  are there versioned releases?  I would prefer that.

File: incubator/openvpn/templates/config-openvpn.yaml [hunk 9]
```diff
@@ -0,0 +1,125 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "fullname" . }}
+data:
+  setup-certs.sh: |-
+    #!/bin/bash
+    apk update && apk add openssl && apk add easy-rsa
```
Comment (coffeepac) [inline incubator/openvpn/templates/config-openvpn.yaml]: is this not included in the source image?  if not, can we make a version that has these installs included?

File: incubator/openvpn/README.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
+# Helm chart for OpenVPN
+This chart will install an openvpn server inside a kubernetes cluster.  New certificates are generated on install, and a script is provided to genreate client keys as needed.  The chart will automatically configure dns to use kubeDNS and route all network traffic to kuberentes pods and services through the vpn.  By connecting to this VPN a host is effectively inside a cluster's network.
+
+###Uses
+The primary purpose of this chart was to make it easy to access kubernetes services during development.  It could also be used for any service that only needs to be access through a VPN, or as a general prupose VPN.
```
Comment (coffeepac) [inline incubator/openvpn/README.md]: 'needs to be accessed' - grammar nit

File: incubator/openvpn/README.md (hunk 3 above)
Comment (coffeepac) [inline incubator/openvpn/README.md]: Hey, this is nonsense. I didn't see this was against a fork of upstream.  Please disregard my comment.

File: incubator/openvpn/README.md (hunk 10 above)
Comment (jfelten) [inline incubator/openvpn/README.md]: I cleaned up many grammar mistakes.

File: incubator/openvpn/Chart.yaml (hunk 1 above)
Comment (jfelten) [inline incubator/openvpn/Chart.yaml]: done.

File: incubator/openvpn/Chart.yaml (hunk 2 above)
Comment (jfelten) [inline incubator/openvpn/Chart.yaml]: added deatil

File: incubator/openvpn/README.md (hunk 3 above)
Comment (jfelten) [inline incubator/openvpn/README.md]: I was planning to add to the general helm repo, since this is unrelated to k2.

**Comment by jfelten** [comment]
Added coffeepac's requested changes

File: incubator/openvpn/templates/NOTES.txt [hunk 11]
```diff
@@ -0,0 +1,20 @@
+OpenVPN is now starting.
```
Comment (coffeepac) [inline incubator/openvpn/templates/NOTES.txt]: I'm kind of confused why this file exists.  Shouldn't any information in this file be in the top level README?  If not, who is the intended audience for this file and when in the chart's lifecycle do you expect to be read.

File: incubator/openvpn/.helmignore [hunk 12]
```diff
@@ -0,0 +1,21 @@
+# Patterns to ignore when building packages.
```
Comment (spiffxp) [inline incubator/openvpn/.helmignore]: Not a blocking comment on this PR, just curious, does helm have the concept of a global helmignore?

eg: I use ~/.gitignore_global to ignore common patterns, so I don't have to put them in each of my projects' .gitignore files

File: incubator/openvpn/Chart.yaml [hunk 13]
```diff
@@ -0,0 +1,18 @@
+apiVersion: extensions/v1beta1
```
Comment (spiffxp) [inline incubator/openvpn/Chart.yaml]: what do we need that forces us to use this instead of v1

File: incubator/openvpn/Chart.yaml [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+description: A Helm chart to install an openvpn server inside a kubernetes cluster.  Certificate generation is also part of the deployment, and this chart will genreate cleint keys as needed.
```
Comment (spiffxp) [inline incubator/openvpn/Chart.yaml]: cleint->client

File: incubator/openvpn/README.md [hunk 15]
```diff
@@ -0,0 +1,47 @@
+# Helm chart for OpenVPN
+This chart will install an openvpn server inside a kubernetes cluster.  New certificates are generated on install, and a script is provided to genreate client keys as needed.  The chart will automatically configure dns to use kubeDNS and route all network traffic to kuberentes pods and services through the vpn.  By connecting to this VPN a host is effectively inside a cluster's network.
```
Comment (spiffxp) [inline incubator/openvpn/README.md]: genreate->generate

kubeDNS->kube-dns

kuberentes->kubernetes

you mix caps on vpn/VPN

what is a host

File: incubator/openvpn/README.md [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+###Uses
+The primary purpose of this chart was to make it easy to access kubernetes services during development.  It could also be used for any service that only needs to be accessed through a VPN or as a standard VPN.
+
+##Usage
+
+		helm repo add incubator http://storage.googleapis.com/kubernetes-charts-incubator
```
Comment (spiffxp) [inline incubator/openvpn/README.md]: this may just be me and markdown etiquette, but I would prefer to see blocks like this surrounded by backticks that call out shell highlighting

File: incubator/openvpn/README.md [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+		helm install incubator/openvpn
+
+Wait for the external load balancer IP to become available.  Check service status via: kubectl get svckubectl get svc
+ 
+Please be aware that certificate generation is variable and may take some time (minutes).
+check pod status via:
```
Comment (spiffxp) [inline incubator/openvpn/README.md]: check->Check

maybe a newline needed here?

File: incubator/openvpn/README.md [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+POD_NAME=`kubectl get pods -l type=openvpn | awk END'{ print $1 }'` \
+&& kubectl log $POD_NAME --follow
+
+When ready generate a client key as follows:
+
+		POD_NAME=`kubectl get pods -l type=openvpn | awk END'{ print $1 }'` \
```
Comment (spiffxp) [inline incubator/openvpn/README.md]: same thing re: backticks

File: incubator/openvpn/README.md [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+Be sure to change KEY_NAME if genreating additional keys.  Import the .ovpn file into your favorite openvpn tool like tunnelblick and verify connectivity.
+
+##Configuration
+
+All configuration is in the values.yaml file, and can be overwritten via the helm --set flag.
+
+###Certificates
+
+Certificates are generated with each deployment.  This way there is no need to store them, but this may be an issue if the pods are restarted often or if there are many clients.  We may enable a persistent storage of certs if this becomes an issue.
```
Comment (spiffxp) [inline incubator/openvpn/README.md]: First time you reference "we".  Can you rephrase to remove?

File: incubator/openvpn/values.yaml [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+  internalPort: 443
+resources:
+  limits:
+    cpu: 300m
+    memory: 128Mi
+  requests:
+    cpu: 300m
+    memory: 128Mi
+openvpn: 
+    #Network allocated for openvpn clients (default: 10.240.0.0).
```
Comment (spiffxp) [inline incubator/openvpn/values.yaml]: consistency: you put comments a space away from hashes above (which I prefer), but have no leading spaces here... which is it

File: incubator/openvpn/templates/_helpers.tpl [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+{{/*
+Expand the name of the chart.
+*/}}
+{{- define "name" -}}
+{{- default .Chart.Name .Values.nameOverride | trunc 24 -}}
+{{- end -}}
+
+{{/*
+Create a default fully qualified app name.
+We truncate at 24 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
```
Comment (spiffxp) [inline incubator/openvpn/templates/_helpers.tpl]: "We" again

File: incubator/openvpn/.helmignore (hunk 12 above)
Comment (jfelten) [inline incubator/openvpn/.helmignore]: yes will add

File: incubator/openvpn/Chart.yaml (hunk 13 above)
Comment (jfelten) [inline incubator/openvpn/Chart.yaml]: There is nothing - I changed it last night while converting and was cutting and pasting.
